src/build_log.h

```diff
@@ -90,7 +106,7 @@ struct BuildLog {
   bool Restat(StringPiece path, const DiskInterface& disk_interface,
               int output_count, char** outputs, std::string* err);
 
-  typedef ExternalStringHashMap<LogEntry*>::Type Entries;
+  typedef ExternalStringHashMap<LogEntryPtr>::Type Entries;
```

Comment:
It *appears* to be the case that this hash map is the "owner" of the log entry.

If that's the case, you could probably use std::unique_ptr, and then return references to that unique ptr where you need to, with the use of std::reference_wrapper<> where appropriate.

Reply:
@jonesmz 

I think it is hard to use `unique_ptr` here because we are using a copy of the pointer in many places:

https://github.com/ninja-build/ninja/blob/1d4034f0ac2b5cfb809b5ab983d47c3cb2c78415/src/build_log.cc#L152-L162

Reply:
Why we cannot get raw pointer from `unique_ptr` there?

Reply:
Ah, totally forgot that, thanks.